python_modules/dagster/dagster/_core/execution/context/compute.py

```diff
@@ -625,6 +626,21 @@ def get_asset_provenance(self, asset_key: AssetKey) -> Optional[DataProvenance]:
             None if record is None else extract_data_provenance_from_entry(record.event_log_entry)
         )
 
+    @public
+    @experimental
+    def report_asset_materialized(
+        self,
+        asset_key: CoercibleToAssetKey,
+        metadata: Optional[Mapping[str, Any]] = None,
+    ):
```

Comment:
what are you thinking in terms of the adding metadata in more than one place in a particular execution? in the way that you can call `add_output_metadata` twice?

Reply:
Also should asset_key be nullable in the case of a single asset getting materialized so we don't have to double encode asset key?

Reply:
> what are you thinking in terms of the adding metadata in more than one place in a particular execution? in the way that you can call add_output_metadata twice?

hmm, we could 
* make `report_` merge metadata on subsequent calls. might get weird when other args are added
* add `add_asset_metadata` that behaves like `add_output_metadata` (but maybe doesn't apply it to the output event as well?)
* something else 

what do you think?